packages/txservice/src/provider.ts

```diff
@@ -0,0 +1,177 @@
+import PriorityQueue from "p-queue";
+import { BaseLogger } from "pino";
+
+import { TransactionServiceConfig } from "./config";
+import { ChainError } from "./error";
+import { FullTransaction, MinimalTransaction, ProviderConfig } from "./types";
+
+const { JsonRpcProvider, FallbackProvider } = providers;
+
+/// Could use a more encompassing name, e.g. ChainRpcDispatch, etc
+export class ChainRpcProvider {
+  private provider: providers.JsonRpcProvider | providers.FallbackProvider;
+  private config: { gasInitialBumpPercent: number; gasPriceMinimum: BigNumber; rpcProviderMaxRetries: number; };
+  private signer: Signer;
+  private queue: PriorityQueue = new PriorityQueue({ concurrency: 1 });
+
+  public confirmationsRequired: number;
+  public chainId: number;
+
+  constructor(
+    private readonly log: BaseLogger,
+    chainId: number,
+    signer: string | Signer,
+    providerConfigs: ProviderConfig[],
+    config: TransactionServiceConfig,
+  ) {
+    const { gasInitialBumpPercent, gasPriceMinimum, rpcProviderMaxRetries } = config;
+    this.config = { gasInitialBumpPercent, gasPriceMinimum, rpcProviderMaxRetries };
+    this.confirmationsRequired = config.chainConfirmations.get(chainId) ?? config.defaultConfirmationsRequired;
+    this.chainId = chainId;
+    
+    // Register a provider for each url.
+    // Make sure all providers are ready()
+
+    if (providerConfigs.length === 1) {
+      this.provider = new JsonRpcProvider(providerConfigs[0]);
+    } else {
+      const fallbackProviderConfigs = providerConfigs.map((config) => ({
+        provider: new JsonRpcProvider(
+          {
+            url: config.url,
+            user: config.user,
+            password: config.password,
+          }
+        ),
+        priority: config.priority ?? 1,
+        weight: config.weight ?? 1,
+        stallTimeout: config.stallTimeout,
+      }));
+      // TODO: Quorum is set to 1 here, but we may want to reconfigure later. Normally it is half the sum of the weights,
+      // which might be okay in our case, but for now we have a low bar.
+      this.provider = new FallbackProvider(fallbackProviderConfigs, 1);
+    }
+
+    this.signer = typeof signer === "string" ? new Wallet(signer, this.provider) : signer.connect(this.provider);
+  }
+
+  public async sendTransaction(
+    tx: FullTransaction
+  ): Promise<{ response: providers.TransactionResponse | Error; success: boolean }> {
+    const ready = await this.provider.ready;
+    if (!ready) {
+      // Error out, no providers are ready.
+      throw new ChainError(ChainError.reasons.ProviderNotReady);
+    }
+    // TODO: Check getBlockNumber() > 0.
+
+    // Define task to send tx with proper nonce
+    const task = async (): Promise<{ response: providers.TransactionResponse | Error; success: boolean }> => {
+      try {
+        // Send transaction using the passed in callback.
+        // const stored = this.nonces.get(chainId);
```

Comment:
We will eventually have to store the nonces in the class to get rid of the race conditions around repeated calls to `pending` and `pending` actually updating

Reply:
Yes we need to store it in the class, but we also need a way to recover from sync issues. One thing we are seeing in Vector is that sometimes the class-stored nonce gets out of sync with chain then it's fully unrecoverable. The only thing that fixes this case is restarting the router which makes it re-sync with chain. In the new system we need to be able to detect and recover from this case. Let's chat more about this @jakekidd.

Also to add on to @LayneHaber's comment, we don't just _eventually_ need this, we need the stateful nonce-tracking now, so may as well implement it.

Reply:
Also it seems like this: https://docs.ethers.io/v5/api/experimental/#experimental-noncemanager

Is purpose-built for tracking nonces like this. Maybe we should just use it?